resolves issues #16 and #17.

miguelcleon opened the pull request:
no more warnings or NOTE messages from R CMD check

miguelcleon commented:
also somewhere the readme was deleted. I added it back in and I made changes to plot.R which you might be working on and could be ignored.

appling commented:
WOW! Very exciting to be green with both builds and on the map with coveralls.

appling commented on R/diagnostics.R:
```diff
@@ -19,6 +19,10 @@
 isTimestepRegular <- function(dates, hist=TRUE, tol=.Machine$double.eps^0.5, handler=stop) {
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  aes=TimeInterval=geom_histogram=xlab=NULL
```
In general for this visible binding error, I've started using ".ggplot2.var", ".dplyr.var", etc. rather than NULL to make it a tiny bit clearer why it needs to be done. That change might be a good idea for TimeInterval - @miguelcleon, would you mind changing it, please?

I'm also pretty sure there's a better fix for aes, geom_histogram, and xlab; could you please try changing the roxygen line to `@importFrom ggplot2 ggplot aes geom_histogram xlab` and removing them from line 24? Those shouldn't ever need to be treated as global variables.

appling replied:
miguel wrote: ok, the import works @importFrom ggplot2 ggplot aes geom_histogram xlab 
I'm not sure where TimeInterval comes from though, I should change it to ??.TimeInterval ?

(if you start a comment with "Add a line note" button, we can maintain several conversation threads in one PR)

appling replied:
great - glad the @importFrom thing helped.

I don't understand your suggestion about ??.TimeInterval ?. Could you paste the full code change you're suggesting?

miguelcleon replied:
Ok, there are two issues at play here. 1) sometimes roxygen doesn't find elements from other packages which you fixed with the @importFrom instead of my =NULL which I agree that makes sense and is better. 2) sometimes roxygen thinks local variables are global variables which is what is happening I believe with s.hat and maybe TimeInterval but I'm not sure if TimeInterval is a local variable or maybe it comes from somewhere else? Then where does it come from? I think the right thing to do in case 2) is still to do the =NULL trick.

appling replied:
i agree that TimeInterval is probably the second type of issue. my best guess without directly experimenting is that it's because of `aes(x=TimeInterval)`, where it's not obvious to R CMD check that `TimeInterval` is a column name within a data.frame. and, yes, let's set it it a value just before the ggplot call. But let's do
```r
TimeInterval  <- '.ggplot.var'
```
instead of NULL, please.

appling commented on R/loadLm.R:
```diff
@@ -88,7 +88,7 @@ loadLm <- function(formula, pred.format=c("flux","conc"),
                    y.trans.function=NULL, retrans.function=exp,
                    store=c("data","fitting.function"), ylog) {
-  
+  s.hat=NULL
```
why?

miguelcleon replied:
It's the same reason I added aes=TimeInterval=geom_histogram=xlab=NULL in the previous issue. I guess roxygen thinks it should be a global variable, seems to be some parsing issue and this circumvents that.

appling replied:
but this is the only line in loadLm where `s.hat` appears.

miguelcleon replied:
line 91 in loadLm.R; I removed this and the note for this section of the code no longer appears.

appling commented on R/loadComp.R:
```diff
@@ -397,7 +397,7 @@ predictSolute.loadComp <- function(
 #' @export
 #' @family estimateMSE
-estimateMSE.loadComp <- function(load.model, n.iter=100, method="parametric", rho) {
+estimateMSE.loadComp <- function(load.model,..., n.iter=100, method="parametric", rho) {
```
glad you were able to eliminate the error, but this strikes me as an odd UI. could you experiment with placing the dots at the end for this and the other estimateMSE definitions? My guess is you'll also have to move all the inconsistent arguments to the end; that's OK with me. Please do a search of estimateMSE through the whole package to fix any places where the function is called and the change in argument orders means that the arguments passed in the function call should also be changed.

miguelcleon replied:
It works to just move ,... to the last argument.

appling replied:
:+1:

miguelcleon commented:
ok, the import works @importFrom ggplot2 ggplot aes geom_histogram xlab 
I'm not sure where TimeInterval comes from though, I should change it to ??.TimeInterval ?

appling commented on R/loadReg2.R:
```diff
@@ -393,7 +393,7 @@ predictSolute.loadReg2 <- function(
 simulateSolute.loadReg2 <- function(load.model, flux.or.conc=c("flux","conc"), newdata, 
                                     method=c("parametric", "non-parametric"), from.interval=c("confidence", "prediction"), rho, ...) {
-  
+  s.hat=NULL
```
again, could you please explain?

appling replied:
as with simulateSolute.lm, i think this is real. please be careful to understand the cause of the error as fully as possible before making corrections - you could have noticed the bug, too. it's a similar fix to that for simulateSolute, but i'll need to think about it some more to get it right. in the meantime, please delete the `s.hat=NULL` line and insert this line just before `noise <- noise * s.hat/sd(noise)`:
```r
error("need to calculate s.hat")
```
then you can comment out the `noise <- noise * s.hat/sd(noise)` line to get rid of the R CMD check error

appling commented on R/loadLm.R:
```diff
@@ -361,7 +361,7 @@ resampleCoefficients.lm <- function(fit) {
 simulateSolute.loadLm <- function(load.model, flux.or.conc=c("flux","conc"), newdata, 
                                   method=c("parametric", "non-parametric"), from.interval=c("confidence", "prediction"), rho, ...) {
-  
+  s.hat=NULL
```
in this case i think R CMD check was catching a real error. please delete line `s.hat=NULL` line and insert this line between the current lines 406 and 407 (just before `noise <- noise * s.hat/sd(noise)`):
```r
s.hat <- summary(load.model@fit)$sigma # residual standard error
```

miguelcleon replied:
line 364 in loadLm.R; I removed this and the note for this section of the code no longer appears.

appling commented on R/metadata.R:
```diff
@@ -287,6 +287,7 @@ metadata <- function(constituent, flow, load.rate="", dates,
 #'   whether \code{new.metadata} is provided.
 #' @export
 updateMetadata <- function(metadata, new.metadata=NA, ..., validate=TRUE) {
+  old=NULL
```
this is the only place in the entire file where `old` occurs. you shouldn't need to do this. could you please try building+checking again with this line deleted, and if you're still getting an error, copy-paste the full error here?

appling commented on R/plots.R:
```diff
@@ -62,7 +62,10 @@ plotObservationsCM <- function(...) {
-  
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  aes_string=geom_point=yday=xlab=NULL
```
same fix here as for `isTimestepRegular`: should be able to put any ggplot2 functions within an `@importFrom ggplot2 ...` line in the roxygen block. `yday` must come from some other package that needs to be imported from in the roxygen, too. `value` and `variable` are the only two that should need to be declared here, and they should please be declared with 
```
value <- variable <- '.ggplot.var'
```
on the line just preceding the ggplot call.

miguelcleon replied:
this is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/plots.R#L66

appling commented on R/plots.R:
```diff
@@ -270,19 +273,22 @@ plotLoadResidualsCM <- function(...){
+plotResidualsCM <- function(type, load.model, observations,resids, dateField="Date", verbose=FALSE,  xObservations=FALSE, xFlow=FALSE, day.of.year=FALSE){
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  aes_string=geom_point=yday=xlab=NULL
```
see previous comment

miguelcleon replied:
this is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/plots.R#L278
and here
https://github.com/miguelcleon/loadflex/blob/master/R/plots.R#L280

appling replied:
pretty sure resids doesn't need to be passed in. but again, plots will be rewritten.

appling commented on R/plots.R:
```diff
@@ -270,19 +273,22 @@ plotLoadResidualsCM <- function(...){
   estName <- ""
   if(type=="conc"){
-    resids <- predictSoluteFromRegression("conc", load.model, observations)
+    #resids <- predictSoluteFromRegression("conc", load.model, observations)
```
this looks problematic to me. not a big deal, since i'll be rewriting this file in a week or two, but any idea why the line was in there in the first place?

miguelcleon replied:
yeah I figured since you are rewriting it. I'd just comment it out. Yeah looks problematic. I'm not really sure.

appling commented on R/plots.R:
```diff
@@ -348,6 +354,7 @@ transformDates <- function(plotsols){
   # Convert to POSIXct if feasible. This will require some thought to figure out
   # which sorts of Period formats can be generated by predLoad and predConc. In
   # the meantime, here we address just a couple of the possibilities.
+  is.POSIXt=DATE=is.Date=NULL
```
definitely not. `is.POSIXt` and `is.Date` are functions that need to be declared with an @importFrom, not by setting them to NULL, this function won't work as changed. DATE should be set like this, please:
```{r}
DATE <- '.transform.var'
```

miguelcleon replied:
This is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/plots.R#L350 
and here
https://github.com/miguelcleon/loadflex/blob/master/R/plots.R#L354

appling commented on R/predictors.R:
```diff
@@ -59,7 +59,9 @@
-  
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  is.Date=is.POSIXt=hysep=NULL
```
definitely not - see previous comment for is.Date and is.POSIXt. For hysep, i actually meant to change this last weekend but seem to have not done that. could you please comment out the entire hysep block in the switch statement, from 71-110? I'd rather make hysep unavailable than commit ourselves to a whole new package dependency just for this one option.

miguelcleon replied:
This is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/predictors.R#L60

appling commented on R/unit.conversions.R:
```diff
@@ -46,6 +46,10 @@ NULL
   # valid.metadata.units
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  numerator=denominator=value=NULL
```
move these lines closer to where the need for them occurs, please. right after line 132, `# Append the reverse conversions`, would be appropriate.

miguelcleon replied:
this is fixed here 
https://github.com/miguelcleon/loadflex/blob/master/R/unit.conversions.R#L130

appling commented on R/unit.conversions.R:
```diff
@@ -164,7 +168,7 @@ generateUnitsData <- function() {
 validMetadataUnits <- function(unitstr, unit.type=c("ANY","flow.units","conc.units","load.units","load.rate.units")) {
   unit.type <- match.arg(unit.type)
-  
+  unit=NULL
```
move closer to where it's needed & use informative character value

miguelcleon replied:
this is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/unit.conversions.R#L178

appling commented on R/unit.conversions.R:
```diff
@@ -217,6 +221,7 @@ validMetadataUnits <- function(unitstr, unit.type=c("ANY","flow.units","conc.uni
 #' loadflex:::translateFreeformToUnitted("mg L^-1") # "mg L^-1"
 translateFreeformToUnitted <- function(freeform.units, attach.units=FALSE) {
   # Quick escape if our work is already done.
+  old=NULL
```
move closer to where it's needed & use informative character value

miguelcleon replied:
this is removed https://github.com/miguelcleon/loadflex/blob/master/R/unit.conversions.R#L178

appling commented on R/unit.conversions.R:
```diff
@@ -273,6 +278,10 @@ flowconcToFluxConversion <- function(flow.units, conc.units, load.rate.units, at
   
+  #silly thing needed to pass R CMD check
+  # from http://stackoverflow.com/questions/8096313/no-visible-binding-for-global-variable-note-in-r-cmd-check
+  numerator=denominator=NULL
```
move closer to where it's needed & use informative character value

miguelcleon replied:
this is fixed here https://github.com/miguelcleon/loadflex/blob/master/R/unit.conversions.R#L290

appling commented on README:
```diff
@@ -0,0 +1,37 @@
+loadflex: Models and Tools for Watershed Flux Estimates
```
so there actually IS a readme, but it's called README.md now (because that gets prettily presented on the GitHub front page for the package). were you seeing errors because of the renaming?

miguelcleon replied:
ok, I just noticed that I had a file called README that wasn't in the repository and I assumed it was removed by mistake. But that sounds good Ill remove it again.

miguelcleon replied:
this is removed 3c217fc6b4c197c7b0e8ec3129b4c85a48d32568

appling commented:
OK, i've read through the full pull request and made some more work for you, @miguelcleon. to help keep track of all that, could you please go through my line notes and respond with both the actual changes and also line notes of your own? i'm going to wait to merge until that's done, so feel free to add more commits to this PR to address the issues raised here.

miguelcleon commented:
Ok, I think that I've addressed all of the above issues in the 2 commits I just pushed. I will double check each line note and add specific responses tomorrow.

miguelcleon commented:
Ok I've responded to all of your line items, let me know if there is anything else for this. Hopefully this is resolved.

appling commented:
comments look good at a glance. i'm headed into a meeting that will probably last all day, but i'll get to these as soon as I can. looks like there might have been some left-over commented-out code...could you clean that up, please, in the meantime?

miguelcleon commented:
yes ok

appling commented on R/plots.R:
```diff
@@ -57,12 +57,11 @@ plotObservationsCM <- function(...) {
                    composite=TRUE, linear.interpolation=FALSE, regression=TRUE,  xrange="none", 
                    dateField="Date", verbose=FALSE) {
-  
+  value <- yday<- variable <- '.ggplot.var'
```
yday is a function, not a variable. i guess we'll ignore this because plots will be rewritten.

appling commented:
looks good.

miguelcleon commented:
great. Let me know if you'd like to assign me something else.
